// @slidev/cli/dist/export-MLH55TH5.mjs
import {
  packageExists,
  resolveGlobalImportPath
} from "./chunk-DWXI5WEO.mjs";
import "./chunk-BXO7ZPPU.mjs";

// node/export.ts
import path from "node:path";
import { Buffer } from "node:buffer";
import fs from "fs-extra";
import { blue, cyan, dim, green, yellow } from "kolorist";
import { Presets, SingleBar } from "cli-progress";
import { parseRangeString } from "@slidev/parser/core";
import { outlinePdfFactory } from "@lillallol/outline-pdf";
import * as pdfLib from "pdf-lib";
import { PDFDocument } from "pdf-lib";
import isInstalledGlobally from "is-installed-globally";
function addToTree(tree, info, slideIndexes, level = 1) {
  const titleLevel = info.level;
  if (titleLevel && titleLevel > level && tree.length > 0) {
    addToTree(tree[tree.length - 1].children, info, slideIndexes, level + 1);
  } else {
    tree.push({
      children: [],
      level,
      path: String(slideIndexes[info.index + 1]),
      hideInToc: Boolean(info.frontmatter?.hideInToc),
      title: info.title
    });
  }
}
function makeOutline(tree) {
  return tree.map(({ title, path: path2, level, children }) => {
    const rootOutline = title ? `${path2}|${"-".repeat(level - 1)}|${title}` : null;
    const childrenOutline = makeOutline(children);
    return childrenOutline.length > 0 ? `${rootOutline}
${childrenOutline}` : rootOutline;
  }).filter((outline) => !!outline).join("\n");
}
function createSlidevProgress(indeterminate = false) {
  function getSpinner(n = 0) {
    return [cyan("\u25CF"), green("\u25C6"), blue("\u25A0"), yellow("\u25B2")][n % 4];
  }
  let current = 0;
  let spinner = 0;
  let timer;
  const progress = new SingleBar({
    clearOnComplete: true,
    hideCursor: true,
    format: `  {spin} ${yellow("rendering")}${indeterminate ? dim(yellow("...")) : " {bar} {value}/{total}"}`,
    linewrap: false,
    barsize: 30
  }, Presets.shades_grey);
  return {
    bar: progress,
    start(total) {
      progress.start(total, 0, { spin: getSpinner(spinner) });
      timer = setInterval(() => {
        spinner += 1;
        progress.update(current, { spin: getSpinner(spinner) });
      }, 200);
    },
    update(v) {
      current = v;
      progress.update(v, { spin: getSpinner(spinner) });
    },
    stop() {
      clearInterval(timer);
      progress.stop();
    }
  };
}
async function exportNotes({
  port = 18724,
  base = "/",
  output = "notes",
  timeout = 3e4
}) {
  const { chromium } = await importPlaywright();
  const browser = await chromium.launch();
  const context = await browser.newContext();
  const page = await context.newPage();
  const progress = createSlidevProgress(true);
  progress.start(1);
  if (!output.endsWith(".pdf"))
    output = `${output}.pdf`;
  await page.goto(`http://localhost:${port}${base}presenter/print`, { waitUntil: "networkidle", timeout });
  await page.waitForLoadState("networkidle");
  await page.emulateMedia({ media: "screen" });
  await page.pdf({
    path: output,
    margin: {
      left: 0,
      top: 0,
      right: 0,
      bottom: 0
    },
    printBackground: true,
    preferCSSPageSize: true
  });
  progress.stop();
  browser.close();
  return output;
}
async function exportSlides({
  port = 18724,
  total = 0,
  range,
  format = "pdf",
  output = "slides",
  slides,
  base = "/",
  timeout = 3e4,
  dark = false,
  routerMode = "history",
  width = 1920,
  height = 1080,
  withClicks = false,
  executablePath = void 0,
  withToc = false,
  perSlide = false,
  scale = 1
}) {
  const pages = parseRangeString(total, range);
  const { chromium } = await importPlaywright();
  const browser = await chromium.launch({
    executablePath
  });
  const context = await browser.newContext({
    viewport: {
      width,
      // Calculate height for every slides to be in the viewport to trigger the rendering of iframes (twitter, youtube...)
      height: perSlide ? height : height * pages.length
    },
    deviceScaleFactor: scale
  });
  const page = await context.newPage();
  const progress = createSlidevProgress(!perSlide);
  async function go(no, clicks) {
    const path2 = `${no}?print${withClicks ? "=clicks" : ""}${clicks ? `&clicks=${clicks}` : ""}${range ? `&range=${range}` : ""}`;
    const url = routerMode === "hash" ? `http://localhost:${port}${base}#${path2}` : `http://localhost:${port}${base}${path2}`;
    await page.goto(url, {
      waitUntil: "networkidle",
      timeout
    });
    await page.waitForLoadState("networkidle");
    await page.emulateMedia({ colorScheme: dark ? "dark" : "light", media: "screen" });
    const elements = page.locator("[data-waitfor]");
    const count = await elements.count();
    for (let index = 0; index < count; index++) {
      const element = elements.nth(index);
      const attribute = await element.getAttribute("data-waitfor");
      if (attribute)
        await element.locator(attribute).waitFor();
    }
    const frames = page.frames();
    await Promise.all(frames.map((frame) => frame.waitForLoadState()));
  }
  async function getSlidesIndex() {
    const clicksBySlide = {};
    const slides2 = page.locator(".print-slide-container");
    const count = await slides2.count();
    for (let i = 0; i < count; i++) {
      const id = await slides2.nth(i).getAttribute("id") || "";
      const path2 = Number(id.split("-")[0]);
      clicksBySlide[path2] = (clicksBySlide[path2] || 0) + 1;
    }
    const slideIndexes = Object.fromEntries(Object.entries(clicksBySlide).reduce((acc, [path2, clicks], i) => {
      acc.push([path2, clicks + (acc[i - 1]?.[1] ?? 0)]);
      return acc;
    }, []));
    return slideIndexes;
  }
  function getClicksFromUrl(url) {
    return url.match(/clicks=([1-9][0-9]*)/)?.[1];
  }
  async function genPageWithClicks(fn, i, clicks) {
    await fn(i, clicks);
    if (withClicks) {
      await page.keyboard.press("ArrowRight", { delay: 100 });
      const _clicks = getClicksFromUrl(page.url());
      if (_clicks && clicks !== _clicks)
        await genPageWithClicks(fn, i, _clicks);
    }
  }
  async function genPagePdfPerSlide() {
    const buffers = [];
    const genPdfBuffer = async (i, clicks) => {
      await go(i, clicks);
      const pdf = await page.pdf({
        width,
        height,
        margin: {
          left: 0,
          top: 0,
          right: 0,
          bottom: 0
        },
        pageRanges: "1",
        printBackground: true,
        preferCSSPageSize: true
      });
      buffers.push(pdf);
    };
    let idx = 0;
    for (const i of pages) {
      await genPageWithClicks(genPdfBuffer, i);
      progress.update(++idx);
    }
    let mergedPdf = await PDFDocument.create({});
    for (const pdfBytes of buffers) {
      const pdf = await PDFDocument.load(pdfBytes);
      const copiedPages = await mergedPdf.copyPages(pdf, pdf.getPageIndices());
      copiedPages.forEach((page2) => {
        mergedPdf.addPage(page2);
      });
    }
    addPdfMetadata(mergedPdf);
    if (withToc)
      mergedPdf = await addTocToPdf(mergedPdf);
    const buffer = await mergedPdf.save();
    await fs.writeFile(output, buffer);
  }
  async function genPagePdfOnePiece() {
    await go("print");
    await page.pdf({
      path: output,
      width,
      height,
      margin: {
        left: 0,
        top: 0,
        right: 0,
        bottom: 0
      },
      printBackground: true,
      preferCSSPageSize: true
    });
    let pdfData = await fs.readFile(output);
    let pdf = await PDFDocument.load(pdfData);
    addPdfMetadata(pdf);
    if (withToc)
      pdf = await addTocToPdf(pdf);
    pdfData = Buffer.from(await pdf.save());
    await fs.writeFile(output, pdfData);
  }
  async function genPagePngOnePiece() {
    await go("print");
    await fs.emptyDir(output);
    const slides2 = await page.locator(".print-slide-container");
    const count = await slides2.count();
    for (let i = 0; i < count; i++) {
      progress.update(i + 1);
      let id = await slides2.nth(i).getAttribute("id") || "";
      id = withClicks ? id : id.split("-")[0];
      const buffer = await slides2.nth(i).screenshot();
      await fs.writeFile(path.join(output, `${id}.png`), buffer);
    }
  }
  async function genPagePngPerSlide() {
    const genScreenshot = async (i, clicks) => {
      await go(i, clicks);
      await page.screenshot({
        omitBackground: false,
        path: path.join(
          output,
          `${i.toString().padStart(2, "0")}${clicks ? `-${clicks}` : ""}.png`
        )
      });
    };
    for (const i of pages)
      await genPageWithClicks(genScreenshot, i);
  }
  function genPagePdf() {
    if (!output.endsWith(".pdf"))
      output = `${output}.pdf`;
    return perSlide ? genPagePdfPerSlide() : genPagePdfOnePiece();
  }
  function genPagePng() {
    return perSlide ? genPagePngPerSlide() : genPagePngOnePiece();
  }
  async function genPageMd(slides2) {
    const files = await fs.readdir(output);
    const mds = files.map((file, i, files2) => {
      const slideIndex = getSlideIndex(file);
      const mdImg = `![${slides2[slideIndex]?.title}](./${path.join(output, file)})

`;
      if ((i + 1 === files2.length || getSlideIndex(files2[i + 1]) !== slideIndex) && slides2[slideIndex]?.note)
        return `${mdImg}${slides2[slideIndex]?.note}

`;
      return mdImg;
    });
    if (!output.endsWith(".md"))
      output = `${output}.md`;
    await fs.writeFile(output, mds.join(""));
  }
  function getSlideIndex(file) {
    const slideId = file.substring(0, file.indexOf(".")).split("-")[0];
    return Number(slideId) - 1;
  }
  function addPdfMetadata(pdf) {
    const titleSlide = slides[0];
    if (titleSlide?.title)
      pdf.setTitle(titleSlide.title);
    if (titleSlide?.frontmatter?.info)
      pdf.setSubject(titleSlide.frontmatter.info);
    if (titleSlide?.frontmatter?.author)
      pdf.setAuthor(titleSlide.frontmatter.author);
    if (titleSlide?.frontmatter?.keywords) {
      if (Array.isArray(titleSlide?.frontmatter?.keywords))
        pdf.setKeywords(titleSlide?.frontmatter?.keywords);
      else
        pdf.setKeywords(titleSlide?.frontmatter?.keywords.split(","));
    }
  }
  async function addTocToPdf(pdf) {
    const outlinePdf = outlinePdfFactory(pdfLib);
    const slideIndexes = await getSlidesIndex();
    const tocTree = slides.filter((slide) => slide.title).reduce((acc, slide) => {
      addToTree(acc, slide, slideIndexes);
      return acc;
    }, []);
    const outline = makeOutline(tocTree);
    return await outlinePdf({ outline, pdf });
  }
  progress.start(pages.length);
  if (format === "pdf") {
    await genPagePdf();
  } else if (format === "png") {
    await genPagePng();
  } else if (format === "md") {
    await genPagePng();
    await genPageMd(slides);
  } else {
    throw new Error(`Unsupported exporting format "${format}"`);
  }
  progress.stop();
  browser.close();
  return output;
}
function getExportOptions(args, options, outDir, outFilename) {
  const config = {
    ...options.data.config.export,
    ...args,
    withClicks: args["with-clicks"],
    executablePath: args["executable-path"],
    withToc: args["with-toc"],
    perSlide: args["per-slide"]
  };
  const {
    entry,
    output,
    format,
    timeout,
    range,
    dark,
    withClicks,
    executablePath,
    withToc,
    perSlide,
    scale
  } = config;
  outFilename = output || options.data.config.exportFilename || outFilename || `${path.basename(entry, ".md")}-export`;
  if (outDir)
    outFilename = path.join(outDir, outFilename);
  return {
    output: outFilename,
    slides: options.data.slides,
    total: options.data.slides.length,
    range,
    format: format || "pdf",
    timeout: timeout ?? 3e4,
    dark: dark || options.data.config.colorSchema === "dark",
    routerMode: options.data.config.routerMode,
    width: options.data.config.canvasWidth,
    height: Math.round(options.data.config.canvasWidth / options.data.config.aspectRatio),
    withClicks: withClicks || false,
    executablePath,
    withToc: withToc || false,
    perSlide: perSlide || false,
    scale: scale || 1
  };
}
async function importPlaywright() {
  if (await packageExists("playwright-chromium"))
    return await import("playwright-chromium");
  let globalPath = isInstalledGlobally ? await resolveGlobalImportPath("playwright-chromium") : void 0;
  if (globalPath)
    return await import(globalPath);
  const { resolveGlobal } = await import("resolve-global");
  try {
    globalPath = resolveGlobal("playwright-chromium");
  } catch {
  }
  if (globalPath)
    return await import(globalPath);
  throw new Error("The exporting for Slidev is powered by Playwright, please install it via `npm i -D playwright-chromium`");
}
export {
  exportNotes,
  exportSlides,
  getExportOptions
};
